File: src/swift/buckets/python/buckets.py

import sys

class NumericalSum(object):
    def call(self, group):
        total = 0 
        for i in range(0, len(group)):
            total += group[i]
        return total

class HistogramReport(object):
    def __init__(self, operation):
        self.operation = operation

    def group_and_process_by_last_digit(self, numbers):
        array = []
        
        array1 = []
        for i in numbers:
            if i < 100:
                array1 += [i]

        digit_groups = [] 

        for i in range(0,10):
            digit_groups += [[]]

        for i in array1:
            last = ("%d" % i)[-1]
            digit_groups[int(last)] += [i]

        for i in range(0,10):
            result = self.operation.call(digit_groups[i])
            array += [result]

        array2 = []
        for i in numbers:
            if i >= 100:
                array2 += [i]

        if len(array2) > 0:
            result = self.operation.call(array2)
            array += [result]

        return array


numbers = []
for line in sys.stdin.readlines():
    for s in line.split():
        numbers += [int(s)]

report = HistogramReport(NumericalSum()).group_and_process_by_last_digit(numbers)
print(report)
    
            

File: src/swift/buckets/java/com/swiftstack/Buckets.java
import java.util.*;

abstract class Callable {
  public abstract Integer call(List group);
}

class NumericalSum extends Callable {
  public NumericalSum() {}
  public Integer call(List group) {
    Integer total = 0;
    for (int i=0; i< group.size(); ++i) {
      Integer element = (Integer) group.get(i);
      total += element.intValue();
    }
    return total;
  }
}

class HistogramReport {
  public Callable operation;
  public HistogramReport(Callable operation) {
    this.operation = operation;
  }

  public List groupAndProcessByLastDigit(List numbers) {
    List array = new ArrayList<Integer>();

    List array1 = new ArrayList<Integer>();
    for (Object number : numbers) {
      int i = (Integer) number;
      if (i < 100) {
	array1.add(i);
      }
    }

    List<Integer>[] digit_groups = new ArrayList[10];

    for (int i=0; i < 10; ++i) {
      digit_groups[i] = new ArrayList<Integer>();
    }

    for (Object element : array1) {
      int i = (Integer) element;
      String last = ("" + i).substring(("" + i).length() - 1);
      digit_groups[Integer.valueOf(last)].add(i);
    }

    for (int i=0; i<10; ++i) {
      Integer result = this.operation.call(digit_groups[i]);
      array.add(result);
    }

    List array2 = new ArrayList<Integer>();
    for (Object number : numbers) {
      int i = ((Integer) number).intValue();
      if (i >= 100) {
	array2.add(i);
      }
    }

    if (array2.size() > 0) {
      Integer result = this.operation.call(array2);
      array.add(result);
    }

    return array;
  }
}

public class Buckets {
  public static void main(String[] argv) {
    List<Integer> numbers = new ArrayList<Integer>();

    Scanner sc = new Scanner(System.in);
    while(sc.hasNextLine()) {
      for (String s : sc.nextLine().split(" +")) {
	numbers.add(Integer.valueOf(s));
      }
    }
    List report = new HistogramReport(new NumericalSum()).
      groupAndProcessByLastDigit(numbers);
    for (Object element : report) {
      Integer intElement = (Integer) element;
      System.out.println(element);
    }
  }
}

